eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs

```diff
@@ -140,264 +438,241 @@ proposalsAddVote voter vote gai (Proposals omap) =
+              )
+
+-- | Get all the descendents of an action-id from the @`Proposals`@ forest
+getAllDescendents ::
+  forall era.
+  Proposals era ->
+  GovActionId (EraCrypto era) ->
+  Set (GovActionId (EraCrypto era))
+getAllDescendents (Proposals omap _roots hierarchy) gai = case OMap.lookup gai omap of
+  Nothing -> Set.empty
+  Just gas -> case gas ^. gasActionL of
+    ParameterChange {} -> collectDescendents pfPParamUpdateL
+    HardForkInitiation {} -> collectDescendents pfHardForkL
+    TreasuryWithdrawals {} -> Set.empty
+    NoConfidence {} -> collectDescendents pfCommitteeL
+    UpdateCommittee {} -> collectDescendents pfCommitteeL
+    NewConstitution {} -> collectDescendents pfConstitutionL
+    InfoAction -> Set.empty
+  where
+    collectDescendents ::
+      Lens' (PForest PGraph era) (PGraph (GovPurposeId p era)) ->
+      Set (GovActionId (EraCrypto era))
+    collectDescendents graphL = Set.map unGovPurposeId $ go graphL $ GovPurposeId gai
+    go ::
+      forall p.
+      Lens' (PForest PGraph era) (PGraph (GovPurposeId p era)) ->
+      GovPurposeId p era ->
+      Set (GovPurposeId p era)
+    go graphL gpi =
+      case Map.lookup gpi $ hierarchy ^. graphL . pGraphNodesL of
+        -- Impossible! getAllDescendents: GovPurposeId not found
+        Nothing -> assert False mempty
+        Just (PNode _parent children) -> children <> foldMap (go graphL) children
+
+-- | Remove the set of given action-ids with their descendents from the
+-- @`Proposals`@ forest
 proposalsRemoveDescendentIds ::
   Set (GovActionId (EraCrypto era)) ->
   Proposals era ->
   (Proposals era, Map (GovActionId (EraCrypto era)) (GovActionState era))
-proposalsRemoveDescendentIds gais ps@(Proposals omap) =
-  proposalsRemoveIds (gais <> foldMap (getAllDescendents omap) gais) ps
+proposalsRemoveDescendentIds gais ps =
+  proposalsRemoveIds (gais <> foldMap (getAllDescendents ps) gais) ps
+
+-- | For use in the @`EPOCH`@ rule. Apply the result of
+-- @`extractDRepPulsingState`@ to the @`Proposals`@ forest, so that:
+--   i. all the expired action-ids and their descendents are removed,
+--   and
+--   ii. the sequence of enacted action-ids is promoted to the root,
+--   removing competing/sibling action-ids and their descendents at each
+--   step
+proposalsApplyEnactment ::
+  forall era.
+  Seq (GovActionId (EraCrypto era)) ->
+  Set (GovActionId (EraCrypto era)) ->
+  Proposals era ->
+  ( Proposals era
+  , Map (GovActionId (EraCrypto era)) (GovActionState era) -- Removed due to enactment
+  , Map (GovActionId (EraCrypto era)) (GovActionState era) -- Removed due to expiry
+  )
+proposalsApplyEnactment enactedGais expiredGais props =
+  let (unexpiredProposals, expiredRemoved) = proposalsRemoveDescendentIds expiredGais props
+      (enactedProposalsState, enactedRemoved) =
+        case foldlM' enact (unexpiredProposals, Map.empty) enactedGais of
+          Nothing -> assert False (props, mempty)
```

Comment:
@lehins do you think we should `proposalsRemoveDescendentIds` here so that we don't ever collect dangling actions in the ledger state?

Reply:
No, `proposalsRemoveIds` is enough, but we do need to make sure that the roots are updated.

Reply:
Just leaving a note: This removal is not needed now since the change to using `GovActionState` for the enacted sequence allowed us to use an `assert` in a single place.